core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,190 @@
+
+  private List<Snapshot> snapshots;
+  private Supplier<List<Snapshot>> snapshotSupplier;
+  private Map<Long, Snapshot> snapshotsById;
+  private Map<String, SnapshotRef> refs;
+
+  SnapshotOperations() {
+    snapshots = Lists.newArrayList();
+    snapshotsById = Maps.newHashMap();
+    refs = Maps.newHashMap();
+  }
+
+  SnapshotOperations(List<Snapshot> snapshots, Map<String, SnapshotRef> refs) {
+    this(snapshots, null, refs);
+  }
+
+  SnapshotOperations(SnapshotOperations snapshotOperations) {
+    this(
+        snapshotOperations.snapshots, snapshotOperations.snapshotSupplier, snapshotOperations.refs);
+  }
+
+  SnapshotOperations(
+      List<Snapshot> snapshots,
+      Supplier<List<Snapshot>> snapshotSupplier,
+      Map<String, SnapshotRef> refs) {
+    this.snapshots = Lists.newArrayList(snapshots);
+    this.snapshotSupplier = snapshotSupplier;
+    this.refs = Maps.newHashMap(refs);
+
+    indexSnapshots();
+  }
+
+  List<Snapshot> snapshots() {
+    ensureLoaded();
+
+    return Collections.unmodifiableList(snapshots);
+  }
+
+  Snapshot snapshot(long id) {
+    if (!snapshotsById.containsKey(id)) {
+      ensureLoaded();
+    }
+
+    return snapshotsById.get(id);
+  }
+
+  boolean contains(long id) {
+    return snapshotsById.containsKey(id);
+  }
+
+  void add(Snapshot snapshot) {
+    ValidationException.check(
+        !contains(snapshot.snapshotId()),
+        "Snapshot already exists for id: %s",
+        snapshot.snapshotId());
+
+    snapshots.add(snapshot);
+    snapshotsById.put(snapshot.snapshotId(), snapshot);
+  }
+
+  List<Snapshot> remove(Collection<Long> idsToRemove) {
+    List<Snapshot> retainedSnapshots =
+        Lists.newArrayListWithExpectedSize(snapshots.size() - idsToRemove.size());
+    List<Snapshot> removedSnapshots = Lists.newArrayListWithExpectedSize(idsToRemove.size());
+
+    for (Snapshot snapshot : snapshots) {
+      long snapshotId = snapshot.snapshotId();
+      if (idsToRemove.contains(snapshotId)) {
+        Snapshot removed = snapshotsById.remove(snapshotId);
+        removedSnapshots.add(removed);
+      } else {
+        retainedSnapshots.add(snapshot);
+      }
+    }
+
+    this.snapshots = retainedSnapshots;
+
+    // remove any refs that are no longer valid
+    Set<String> danglingRefs = Sets.newHashSet();
+    for (Map.Entry<String, SnapshotRef> refEntry : refs.entrySet()) {
+      if (!snapshotsById.containsKey(refEntry.getValue().snapshotId())) {
+        danglingRefs.add(refEntry.getKey());
+      }
+    }
+
+    danglingRefs.forEach(this::removeRef);
+
+    return removedSnapshots;
+  }
+
+  Map<String, SnapshotRef> refs() {
+    return Collections.unmodifiableMap(refs);
+  }
+
+  SnapshotRef ref(String name) {
+    return refs.get(name);
+  }
+
+  void addRef(String name, SnapshotRef ref) {
+    refs.put(name, ref);
+  }
+
+  SnapshotRef removeRef(String name) {
+    return refs.remove(name);
+  }
+
+  void indexSnapshots() {
+    this.snapshotsById =
```

Comment:
It is not very obvious how this field is initialized as it is somewhere deep in private methods. I'd consider a similar strategy we do in `TableMetadata`, for example, where we have a method that can index something but the return value of that method is explicitly assigned.

```
this.snapshotsById = indexSnapshots(snapshots);
```

```
private Map<Long, Snapshot> indexSnapshots(List<Snapshot> snapshotsAsList) {
  ...
}
```

Reply:
Updated and cleaned up the naming.